[VPlan] Remove loop region in optimizeForVFAndUF.

fhahn opened the pull request:
Update optimizeForVFAndUF to completely remove the vector loop region when possible. At the moment, we cannot remove the region if it contains

 * widened IVs: the recipe is needed to generate the step vector
 * reductions: ComputeReductionResults requires the reduction phi recipe for codegen.

Both cases can be addressed by more explicit modeling.

The patch also includes a number of updates to allow executing VPlans without a vector loop region which can be split off.

Depends on https://github.com/llvm/llvm-project/pull/110004

fhahn committed:
[VPlan] Remove loop region in optimizeForVFAndUF.

fhahn committed:
[VPlan] Use ResumePhi to create reduction resume phis.

Use VPInstruction::ResumePhi to create phi nodes for reduction resume
values.

This allows simplifying createAndCollectMergePhiForReduction to only
collect reduction resume phis when vectorizing epilogue loops and adding
extra incoming edges from the main vector loop.

fhahn commented:
This now depends https://github.com/llvm/llvm-project/pull/110004 and contains changes from it

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup after merge

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7632,24 +7621,26 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
+    Loop *L = LI->getLoopFor(State.CFG.VPBB2IRBB[HeaderVPBB]);
+    if (VectorizedLoopID)
+      L->setLoopID(*VectorizedLoopID);
+    else {
```
Added, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -696,16 +696,46 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
-
   SmallVector<VPValue *> PossiblyDead(Term->operands());
   Term->eraseFromParent();
+  VPBasicBlock *Header =
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -696,16 +696,46 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+
+    VPBlockBase *Preheader = Plan.getVectorLoopRegion()->getSinglePredecessor();
+    auto HeaderSuccs = to_vector(Header->getSuccessors());
+    VPBasicBlock *Exiting =
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -714,7 +744,7 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
 
 /// Sink users of \p FOR after the recipe defining the previous value \p
-/// Previous of the recurrence. \returns true if all users of \p FOR could be
+// Previous of the recurrence. \returns true if all users of \p FOR could be
```
Removed change, thanks!

fhahn committed:
!fixup address latest comments, thanks!

fhahn reviewed (commented):
Rebased after landing https://github.com/llvm/llvm-project/pull/110004

ping :)

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn commented:
ping :)

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1042,53 +1041,59 @@ void VPlan::execute(VPTransformState *State) {
-      // consistent placement of all induction updates.
-      Instruction *Inc = cast<Instruction>(Phi->getIncomingValue(1));
-      Inc->moveBefore(VectorLatchBB->getTerminator()->getPrevNode());
+      if (isa<VPWidenPointerInductionRecipe>(&R) ||
+          isa<VPWidenIntOrFpInductionRecipe>(&R)) {
```
```suggestion
      if (isa<VPWidenPointerInductionRecipe, VPWidenIntOrFpInductionRecipe>(&R)) {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1042,53 +1041,59 @@ void VPlan::execute(VPTransformState *State) {
+      auto *PhiR = cast<VPHeaderPHIRecipe>(&R);
+      bool NeedsScalar =
+          isa<VPCanonicalIVPHIRecipe, VPEVLBasedIVPHIRecipe>(PhiR) ||
+          (isa<VPReductionPHIRecipe>(PhiR) &&
+           cast<VPReductionPHIRecipe>(PhiR)->isInLoop());
```
Replace isa+cast by dyn_cast?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -692,16 +692,46 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
   SmallVector<VPValue *> PossiblyDead(Term->operands());
   Term->eraseFromParent();
+  auto *Header = cast<VPBasicBlock>(Plan.getVectorLoopRegion()->getEntry());
+  if (all_of(Header->phis(), [](VPRecipeBase &R) {
+        return !isa<VPWidenIntOrFpInductionRecipe, VPReductionPHIRecipe>(&R);
+      })) {
```
```suggestion
  if (none_of(Header->phis(), IsaPred<VPWidenIntOrFpInductionRecipe, VPReductionPHIRecipe>)) {
```

fhahn replied:
Thanks, updated and reordered conditions

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -692,16 +692,46 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+    VPBlockUtils::connectBlocks(Exiting, Middle);
+    // Set LoopRegion's Entry to nullptr, as the CFG from LoopRegion shouldn't
+    // be deleted when the region is deleted.
+    LoopRegion->clearEntry();
+    delete LoopRegion;
```
Potentially unsafe, may leak if the programmer forgets about it. Maybe, use Destructor or something like RAII?

fhahn replied:
I'm not sure if there is a nice way to do so currently. The whole replace-and-erase logic in the block needs to happen atomically at the moment. To avoid having to do manual delete's, VPlan could track all blocks added to it and delete them all when the VPlan is deleted (at the moment it only deletes reachable blocks). But that would require some extra logic to make sure all blocks are added properly.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
!fixup fix formatting

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup update after merge.

fhahn committed:
!fixup remove unrelated changes.

fhahn commented:
Updated to current `main`, ping :)

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup partial fixup after merging main.

alexey-bataev reviewed (approved):
LG

fhahn commented:
After merging initial skeleton creation, we cannot easily get the vector preheader any longer without the region. Putting this on hold for a bit.

fhahn commented:
> LG

Unfortunately on latest main we cannot easily identify the vector preheader when constructing the skeleton if the vector loop region is removed :(  Need to think if there is a nice solution

fhahn committed:
[VPlan] Manage created blocks directly in VPlan. (NFC)

This patch changes the way blocks are managed by VPlan. Previously all
blocks reachable from entry would be cleaned up when a VPlan is
destroyed. With this patch, each VPlan keeps track of blocks created for
it in a list and this list is then used to delete all blocks in the list
when the VPlan is destroyed. To do so, block creation is funneled
through helpers in directly in VPlan.

The main advantage of doing so is it simplifies CFG transformations, as
those do not have to take care of deleting any blocks, just adjusting
the CFG. This helps to simplify
https://github.com/llvm/llvm-project/pull/108378 and
https://github.com/llvm/llvm-project/pull/106748.

This also simplifies handling of 'immutable' blocks a VPlan holds
references to, which at the moment only include the scalar
header block.

Note that the original constructors taking VPBlockBase are retained at
the moment for unit tests.

fhahn commented:
It turns out that tracking the blocks to delete in VPlan directly would help to simply other transforms as well (https://github.com/llvm/llvm-project/pull/106748 in particular) so I put up https://github.com/llvm/llvm-project/pull/120918

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup update on top of current main, iterate to find vector.ph

fhahn committed:
!fixup address comments, add comments

fhahn committed:
[VPlan] Funnel

fhahn committed:
[VPlan] Funnel

fhahn committed:
Merge remote-tracking branch 'origin/main' into manage-blocks-in-vplan

fhahn committed:
Merge remote-tracking branch 'origin/main' into manage-blocks-in-vplan

fhahn committed:
Merge branch 'manage-blocks-in-vplan' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup update after merge

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup adjust assert

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup update extra tests.

fhahn committed:
!fixup fix formatting

fhahn committed:
[VPlan] Replace VPBBs with VPIRBBs during skeleton creation (NFC).

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
[VPlan] Track VectorPH during skeleton creation.

fhahn committed:
!fixup restore original getVectorLoopRegion impl

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2446,9 +2450,11 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
+                                       BasicBlock *CheckIRBB) {
   VPBlockBase *ScalarPH = Plan.getScalarPreheader();
-  VPBlockBase *VectorPH = Plan.getVectorPreheader();
+  // FIXME: Cannot get the vector preheader at the moment if the vector loop
+  // region has been removed.
```
Passing in `VectorPH` seems fine, still need "FIXME"? In fact, can pass both `VectorDestination` and `ScalarDestination`, i.e., the two block that the new check block is introduced to select between.

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -676,6 +678,8 @@ class InnerLoopVectorizer {
   BasicBlock *AdditionalBypassBlock = nullptr;
 
   VPlan &Plan;
+
+  VPBasicBlock *VectorPHVPBB;
```
Follow-up: when skeleton creation moves to VPlan creation, ILV should not depend on VPlan or its VPBB's, as was originally the case.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3668,8 +3668,6 @@ class VPRegionBlock : public VPBlockBase {
   const VPBlockBase *getEntry() const { return Entry; }
   VPBlockBase *getEntry() { return Entry; }
 
-  /// Set \p EntryBlock as the entry VPBlockBase of this VPRegionBlock. \p
-  /// EntryBlock must have no predecessors.
```
Documentation dropped?

fhahn replied:
Dropped by accident, added back, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -794,11 +794,14 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
 
-/// Try to simplify the recipes in \p Plan
-static void simplifyRecipes(VPlan &Plan) {
+/// Try to simplify the recipes in \p Plan. If \p CanonicalIVTy is not nullptr,
+/// use it directly instead of retrieving the canonical IV type from the plan
+/// which may not exist any longer.
```
Simpler to always pass in this Type, which is essentially the type of all "symbolic live-ins" - those whose underlying IR value will be generated later / at code-gen. Or find a way for each such live-in to provide its type.

fhahn replied:
Done thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3846,7 +3844,10 @@ class VPlan {
 
   /// Returns the preheader of the vector loop region.
```
```suggestion
  /// Returns the preheader of the vector loop region, if one exists, or null otherwise.
```

fhahn replied:
Updated, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -837,16 +840,42 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
       !SE.isKnownPredicate(CmpInst::ICMP_ULE, TripCount, C))
     return;
 
```
Say something about what is about to happen now.

fhahn replied:
Added, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -837,16 +840,42 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+    simplifyRecipes(Plan, CanIVTy);
+  }
+
+  VPlanTransforms::removeDeadRecipes(Plan);
```
Worth committing this replacement of recursivelyDeleteDeadRecipes() separately?

fhahn replied:
Done, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -837,16 +840,42 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+        VPInstruction::BranchOnCond,
+        {Plan.getOrAddLiveIn(ConstantInt::getTrue(Ctx))}, Term->getDebugLoc());
+    ExitingVPBB->appendRecipe(BOC);
+  } else {
```
Say something about what is about to happen now.

fhahn replied:
Added, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -837,16 +840,42 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+    ExitingVPBB->appendRecipe(BOC);
+  } else {
+    for (VPRecipeBase &R : make_early_inc_range(Header->phis())) {
+      auto *P = cast<VPHeaderPHIRecipe>(&R);
```
```suggestion
      auto *HeaderPhiR = cast<VPHeaderPHIRecipe>(&R);
```

fhahn replied:
Done thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -837,16 +840,42 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+      P->eraseFromParent();
+    }
+
+    VPBlockBase *Preheader = Plan.getVectorPreheader();
+    VPBlockBase *Middle = Plan.getMiddleBlock();
```
Better take VectorRegion's single predecessor and single successor?

fhahn replied:
Done thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -837,16 +840,42 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
+
+    for (VPBlockBase *B : vp_depth_first_shallow(VectorRegion->getEntry())) {
+      if (isa<VPRegionBlock>(B))
+        B->setParent(nullptr);
```
What about basic block B's, between replicate regions - other than Header and ExitingVPBB?

fhahn replied:
Updated to clear parent for all blocks and removed setting just header and exitingVPBB above, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -936,7 +935,6 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
 
   IRBuilder<> Builder(State.CFG.PrevBB->getTerminator());
   // FIXME: Model VF * UF computation completely in VPlan.
-  assert(VFxUF.getNumUsers() && "VFxUF expected to always have users");
```
We may remove the users (the canonical IV increment). Updated to account for that in assertion

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -487,7 +486,7 @@ void VPBasicBlock::execute(VPTransformState *State) {
 
   // 1. Create an IR basic block.
-  if (this == getPlan()->getVectorPreheader() ||
+  if (this == getPlan()->getEntry() ||
```
Actually this check can be removed completely, as the pre-header and entry are modeled as VPIRBasicBlocks.

Split off to 20d491bb993218eae6a13e4901da943ebd804f7a

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1015,49 +1013,51 @@ void VPlan::execute(VPTransformState *State) {
-        auto *GEP = cast<GetElementPtrInst>(State->get(WidenPhi));
-        Phi = cast<PHINode>(GEP->getPointerOperand());
-      }
+  if (auto *LoopRegion = getVectorLoopRegion()) {
```
Updated, thanks! DTU.flush can be pulled up

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1427,8 +1436,10 @@ void VPlanIngredient::print(raw_ostream &O) const {
   return !hasDefiningRecipe() ||
-         !getDefiningRecipe()->getParent()->getEnclosingLoopRegion();
+         (!getDefiningRecipe()->getParent()->getEnclosingLoopRegion() &&
+          getDefiningRecipe()->getParent()->getPlan()->getVectorLoopRegion());
```
Done with a comment, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2376,7 +2376,9 @@ void VPBranchOnMaskRecipe::execute(VPTransformState &State) {
-  assert(isa<UnreachableInst>(CurrentTerminator) &&
+  assert((isa<UnreachableInst>(CurrentTerminator) ||
+          (isa<BranchInst>(CurrentTerminator) &&
+           !CurrentTerminator->getOperand(0))) &&
```
Not needed with the latest version, removed, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -557,7 +556,7 @@ VPBasicBlock *VPBasicBlock::splitAt(iterator SplitAt) {
   if (P && P->isReplicator()) {
     P = P->getParent();
-    assert(!cast<VPRegionBlock>(P)->isReplicator() &&
+    assert((!P || !cast<VPRegionBlock>(P)->isReplicator()) &&
```
Added a comment but left the check as is for now.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2383,12 +2383,12 @@ void InnerLoopVectorizer::scalarizeInstruction(const Instruction *Instr,
-                           [](VPValue *Op) {
-                             return Op->isDefinedOutsideLoopRegions();
-                           })) &&
-         "Expected a recipe is either within a region or all of its operands "
-         "are defined outside the vectorized region.");
+  assert(
+      (Parent || !RepRecipe->getParent()->getPlan()->getVectorLoopRegion() ||
+       all_of(RepRecipe->operands(),
+              [](VPValue *Op) { return Op->isDefinedOutsideLoopRegions(); })) &&
+      "Expected a recipe is either within a region or all of its operands "
+      "are defined outside the vectorized region.");
   if (IfPredicateInstr)
```
Yes but such a helper needs to be added. Will check if there are other users that could benefit, thanks

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/call-costs.ll:
```diff
@@ -80,15 +80,13 @@ define void @powi_call(ptr %P) {
-; CHECK-NEXT:    store <2 x double> [[TMP3]], ptr [[TMP2]], align 8
-; CHECK-NEXT:    [[INDEX_NEXT]] = add nuw i64 [[INDEX]], 2
-; CHECK-NEXT:    br i1 true, label %[[MIDDLE_BLOCK:.*]], label %[[VECTOR_BODY]], !llvm.loop [[LOOP4:![0-9]+]]
+; CHECK-NEXT:    [[TMP4:%.*]] = getelementptr inbounds double, ptr [[TMP1]], i32 0
```
Yes, this is a fall-out from not running non-VPlan-based simple CSE in fixVectorizedLoop

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/call-costs.ll:
```diff
@@ -80,15 +80,13 @@ define void @powi_call(ptr %P) {
-; CHECK-NEXT:    br i1 true, label %[[MIDDLE_BLOCK:.*]], label %[[VECTOR_BODY]], !llvm.loop [[LOOP4:![0-9]+]]
+; CHECK-NEXT:    [[TMP4:%.*]] = getelementptr inbounds double, ptr [[TMP1]], i32 0
+; CHECK-NEXT:    store <2 x double> [[TMP3]], ptr [[TMP4]], align 8
+; CHECK-NEXT:    br label %[[MIDDLE_BLOCK:.*]]
```
Yep

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
!fixup fix formatting

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-loop-region-instead-of-using-branch-on-cond-true

fhahn committed:
!fixup restore newline and move comment

fhahn committed:
!fixup move erase of term just before removing dead recipes.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1399,11 +1406,17 @@ void VPlanIngredient::print(raw_ostream &O) const {
 
+bool VPValue::isDefinedOutsideLoopRegions() const {
+  return !isDefinedInsideLoopRegions(this);
+}
```
post-commit nit:
```suggestion
}

```
